### PR TITLE
refactor data transformation

### PR DESCRIPTION
**Pull request by aleks-luk**

### CODE REVIEW

**Comment by aleks-luk** [comment]
Grzegorz, my thoughts:

-  I think we should we should delete function save to csv from scrapper_books.py in order to have transformation and saving files in one place. 
- Same thing with api_random_users.py during the refactoring code
- Overall I believe I should have keep transformations, saving in one file. Other files should just pull data from sources

File: books/data_transformation.py [hunk 1]
```diff
@@ -1,108 +1,128 @@
...
 if __name__ == '__main__':
-    transform_books()
-    transform_users()
-    transform_ratings()
-    transform_user_api_data()
-    transform_scrapped_books_data()
+    parser = ArgumentParser(prog='Data Transformation')
+    parser.add_argument('-t', '--target_path', default='./_data/books/transformed', type=str)
+    parser.add_argument('-s', '--source_path', default='./_data/books/raw', type=str)
+    args = parser.parse_args()
```
Comment (kraszewskigrzeg) [inline books/data_transformation.py:117]: This is fine!

File: books/data_transformation.py [hunk 2]
```diff
@@ -1,108 +1,128 @@
...
+    transformation.transform_books(source_path=args.source_path)
+    transformation.transform_users(source_path=args.source_path)
+    transformation.transform_ratings(source_path=args.source_path)
+    # transformation.transform_user_api_data(source_path=args.source_path)
+    # transformation.transform_scrapped_books_data(source_path=args.source_path) >I think we should delete fun save to csv from scrapper_books.py in order to have transformation and saving files in one place
+    transformation.save_to_csv('books', target_path=args.target_path)
+    transformation.save_to_csv('users', target_path=args.target_path)
+    transformation.save_to_csv('ratings', target_path=args.target_path)
+    # transformation.save_to_csv('api_users', target_path=args.target_path)
+    # python books/data_transformation.py -t './_data/books/transformed' -s './_data/books/raw'
```
Comment (kraszewskigrzeg) [inline books/data_transformation.py:128]: This is fine but..... If I were writing this script I would do it in a way that:
```python
transformation = DataTransformation()
transformation.transform_books(source_path=args.source_path, target_path=args.target_path)
```

and call only a single method to transform data :)

File: books/data_transformation.py [hunk 3]
```diff
@@ -1,108 +1,128 @@
...
+            'Title': 'book_title',
+            'Author': 'book_author',
+            'Rok wydania:': 'year_published',
+            'Wydawnictwo:': 'publisher',
+        }
+        scrapped_data.rename(columns=new_scraped_data_column_names, inplace=True)
+        scrapped_data['year_published'] = scrapped_data['year_published'].fillna(0).astype(int)
+        self.data_frames['scrapped_data'] = scrapped_data
+
+    def save_to_csv(self, data_type, target_path):
```
Comment (kraszewskigrzeg) [inline books/data_transformation.py:103]: `data_type` is very missleading here :/

Comment (kraszewskigrzeg) [reply books/data_transformation.py:103]: Generally this method could be moved outside of a class as it shoudl only save a dataframe nothing else.

File: books/data_transformation.py [hunk 4]
```diff
@@ -1,108 +1,128 @@
...
+        ])
+
+        new_api_users_column_names = {
+            'location.city': 'user_city',
+            'location.country': 'user_country',
+            'registered.age': 'user_age'
+        }
+        df_users.rename(columns=new_api_users_column_names, inplace=True)
+        df_users['user_age'] = df_users['user_age'].fillna(0).astype(int)
+        self.data_frames['api_users'] = df_users
```
Comment (kraszewskigrzeg) [inline books/data_transformation.py:88]: this should not be here at all. Just save the results to the target file here :)

File: books/data_transformation.py [hunk 5]
```diff
@@ -1,108 +1,128 @@
...
+        self.data_frames['users'] = users_frame
+
+    def transform_ratings(self, source_path):
+        ratings_frame = pd.read_csv(os.path.join(source_path, 'ratings.csv'), usecols=['ISBN', 'Book-Rating'], delimiter=';', encoding='windows-1250', low_memory=False)
+        new_ratings_column_names = {
+            'ISBN': 'r_book_isbn',
+            'Book-Rating': 'r_book_rating',
+        }
+        ratings_frame.rename(columns=new_ratings_column_names, inplace=True)
+        self.data_frames['ratings'] = ratings_frame
```
Comment (kraszewskigrzeg) [inline books/data_transformation.py:61]: this should not be here at all. Just save the results to the target file here :)

File: books/data_transformation.py [hunk 6]
```diff
@@ -1,108 +1,128 @@
...
+            'Age': 'user_age',
+            0: 'user_city',
+            2: 'user_country'
+        }
+
+        users_frame.rename(columns=new_users_column_names, inplace=True)
+        users_frame['user_city'] = users_frame['user_city'].str.title()
+        users_frame['user_country'] = users_frame['user_country'].str.title()
+        users_frame['user_age'] = users_frame['user_age'].fillna(0).astype(int)
+        self.data_frames['users'] = users_frame
```
Comment (kraszewskigrzeg) [inline books/data_transformation.py:52]: this should not be here at all. Just save the results to the target file here :)

File: books/data_transformation.py [hunk 7]
```diff
@@ -1,108 +1,128 @@
...
+        new_books_column_names = {
+            'ISBN': 'book_isbn',
+            'Book-Title': 'book_title',
+            'Book-Author': 'book_author',
+            'Year-Of-Publication': 'year_published',
+            'Publisher': 'publisher',
+            'Image-URL-S': 'book_image_url'
+        }
+        books_frame.rename(columns=new_books_column_names, inplace=True)
+        self.data_frames['books'] = books_frame
```
Comment (kraszewskigrzeg) [inline books/data_transformation.py:33]: this should not be here at all. Just save the results to the target file here :)

**Review by kraszewskigrzeg — CHANGES_REQUESTED** [review]
Additional questions:
- should the api users be refactored too?
- did you merged main into current branch to assert all changes are there? (this is a little more advence and might not be applicable)

File: books/data_transformation.py (hunk 2 above)
Comment (kraszewskigrzeg) [reply books/data_transformation.py:128]: Yes exactly!

File: books/data_transformation.py [hunk 8]
```diff
@@ -1,108 +1,128 @@
...
+
+
+class DataTransformation:
+
+    def __init__(self):
+        self.data_frames = {}
+        logging.info("DataTransformation initialized")
+
+    def transform_books(self, source_path):
+        books_frame = pd.read_csv(os.path.join(source_path, 'books.csv'), delimiter=';',
```
[the remark this answers is not in the document]
Comment (kraszewskigrzeg) [reply books/data_transformation.py:20]: This to be discussed live!

**Comment by kraszewskigrzeg** [comment]
I have resolved conflicts be sure to pull repo at your end before next changes :)

**Review by kraszewskigrzeg — APPROVED** [review]

**Pull request merged**